### PR TITLE
Add receive, send listener for getting performance metrics

### PR DESCRIPTION
**Pull request by tilumi**
Allow user can set listeners to retrieve EventHub receiving, sending performance metrics.
Receiver Listener interface: `EventHubsReceiverListener`
Sender Listener interface: `EventHubsSenderListener`
During setting listener, the listener is serialized to base 64 string by `SerializationUtils` & during getting, the listener is deserialized by `SerializationUtils`

### CODE REVIEW

File: core/src/main/scala/org/apache/spark/eventhubs/ConnectionStringBuilder.scala [hunk 1]
```diff
@@ -240,6 +240,10 @@ class ConnectionStringBuilder private () {
     this
   }
 
+  def getNamespace: String = {
+    this.getEndpoint.getHost.dropRight(".servicebus.windows.net".length)
```
Comment (nyaghma) [inline core/src/main/scala/org/apache/spark/eventhubs/ConnectionStringBuilder.scala]: ".servicebus.windows.net" is not the only available domain. The assumption that all namespaces are in this domain is incorrect. Also, no value should be hardcoded in a method.

File: core/src/main/scala/org/apache/spark/eventhubs/client/CachedEventHubsReceiver.scala [hunk 2]
```diff
@@ -274,23 +277,27 @@ private[spark] object CachedEventHubsReceiver extends CachedReceiver with Loggin
...
     }
     receiver.receive(requestSeqNo, batchSize)
   }
 
   def apply(ehConf: EventHubsConf,
             nAndP: NameAndPartition,
-            startSeqNo: SequenceNumber): CachedEventHubsReceiver = {
-    new CachedEventHubsReceiver(ehConf, nAndP, startSeqNo)
+            startSeqNo: SequenceNumber,
+            eventHubsReceiverListener: Option[EventHubsReceiverListener]): CachedEventHubsReceiver = {
```
Comment (nyaghma) [inline core/src/main/scala/org/apache/spark/eventhubs/client/CachedEventHubsReceiver.scala]: eventHubsReceiverListener should have a default None value to ensure backward compatibility

File: core/src/test/scala/org/apache/spark/eventhubs/EventHubsConfSuite.scala [hunk 3]
```diff
@@ -207,6 +208,46 @@ class EventHubsConfSuite extends FunSuite with BeforeAndAfterAll {
     assert(actual.equals(expected))
   }
 
+  test("receiverListener serialization / deserialization") {
```
Comment (nyaghma) [inline core/src/test/scala/org/apache/spark/eventhubs/EventHubsConfSuite.scala]: More test cases are needed. Current test cases cover serialization/deserialization for both the sender and receiver but do not test the correctness of the collected metrics. At least two more test cases (one for the sender, one for the receiver) should be added to test the correctness of the collected metrics.

Comment (tilumi) [reply core/src/test/scala/org/apache/spark/eventhubs/EventHubsConfSuite.scala]: How to test correctness of collected metrics? we don't have a mocked eventHub server to test against, or I can only generate some test metrics, pass in the metric to the listener, and verify the value in listener are equal to the passed-in value?

Comment (nyaghma) [reply core/src/test/scala/org/apache/spark/eventhubs/EventHubsConfSuite.scala]: You can use the EventHubsTestUtils class to create a simulated, in-memory Event Hubs instance. You can send a fixed number/size of events to the simulated instance, then receive those events and make sure the collected metrics during both send/receive are as expected.

Comment (tilumi) [reply core/src/test/scala/org/apache/spark/eventhubs/EventHubsConfSuite.scala]: OK, thanks!

File: core/src/main/scala/org/apache/spark/eventhubs/utils/EventHubsReceiverListener.scala [hunk 4]
```diff
@@ -0,0 +1,14 @@
+package org.apache.spark.eventhubs.utils
+
+import com.microsoft.azure.eventhubs.EventData
+import org.apache.spark.eventhubs.{NameAndPartition, SequenceNumber}
+
+trait EventHubsReceiverListener extends Serializable {
+
+  def onBatchReceiveSuccess(nAndP: NameAndPartition, elapsedTime: Long, batchSize: Int, receivedBytes: Long): Unit
+
+  def onBatchReceiveSkip(nAndP: NameAndPartition, requestSeqNo: SequenceNumber, batchSize: Int): Unit
```
Comment (tilumi) [inline core/src/main/scala/org/apache/spark/eventhubs/utils/EventHubsReceiverListener.scala:10]: will remove it

File: core/src/main/scala/org/apache/spark/eventhubs/utils/EventHubsReceiverListener.scala [hunk 5]
```diff
@@ -0,0 +1,14 @@
...
+
+  def onReceiveFirstEvent(nAndP: NameAndPartition, firstEvent: EventData): Unit
```
Comment (tilumi) [inline core/src/main/scala/org/apache/spark/eventhubs/utils/EventHubsReceiverListener.scala:12]: will remove.

File: core/src/main/scala/org/apache/spark/eventhubs/utils/EventHubsSenderListener.scala [hunk 6]
```diff
@@ -0,0 +1,15 @@
+package org.apache.spark.eventhubs.utils
+
+trait EventHubsSenderListener extends Serializable {
+
+  def onBatchSendSuccess(messageCount: Int, messageSizeInBytes: Int, sendElapsedTimeInNanos: Long, retryTimes: Int)
+
+  def onBatchSendFail(exception: Throwable)
```
Comment (tilumi) [inline core/src/main/scala/org/apache/spark/eventhubs/utils/EventHubsSenderListener.scala:7]: Will remove

File: core/src/main/scala/org/apache/spark/eventhubs/client/CachedEventHubsReceiver.scala [hunk 7]
```diff
@@ -211,6 +208,7 @@ private[client] class CachedEventHubsReceiver private (ehConf: EventHubsConf,
     // Retrieve the events. First, we get the first event in the batch.
     // Then, if the succeeds, we collect the rest of the data.
     val first = Await.result(checkCursor(requestSeqNo), ehConf.internalOperationTimeout)
+    eventHubsReceiverListener.foreach(_.onReceiveFirstEvent(nAndP, first.head))
```
Comment (tilumi) [inline core/src/main/scala/org/apache/spark/eventhubs/client/CachedEventHubsReceiver.scala:211]: will remove

File: core/src/main/scala/org/apache/spark/eventhubs/package.scala [hunk 8]
```diff
@@ -49,6 +52,8 @@ package object eventhubs {
   val DefaultThreadPoolSize = 16
   val DefaultEpoch = 0L
   val RetryCount = 3
+  val InternalOperationTimeout: FiniteDuration = 300.seconds
```
Comment (tilumi) [inline core/src/main/scala/org/apache/spark/eventhubs/package.scala:55]: will remove

File: core/src/main/scala/org/apache/spark/eventhubs/package.scala [hunk 9]
```diff
@@ -49,6 +52,8 @@ package object eventhubs {
...
+  val DefaultReceiveRetryTimes: Int = 3
```
Comment (tilumi) [inline core/src/main/scala/org/apache/spark/eventhubs/package.scala:56]: will remove

File: core/src/main/scala/org/apache/spark/eventhubs/client/CachedEventHubsReceiver.scala [hunk 10]
```diff
@@ -59,11 +55,12 @@ private[spark] trait CachedReceiver {
...
  * @param ehConf the [[EventHubsConf]] which contains the connection string used to connect to Event Hubs
  * @param nAndP  the Event Hub name and partition that the receiver is connected to.
- */
+*/
 private[client] class CachedEventHubsReceiver private (ehConf: EventHubsConf,
                                                        nAndP: NameAndPartition,
-                                                       startSeqNo: SequenceNumber)
-    extends Logging {
+                                                       startSeqNo: SequenceNumber,
+                                                       eventHubsReceiverListener: Option[MetricPlugin])
```
Comment (sjkwak) [inline core/src/main/scala/org/apache/spark/eventhubs/client/CachedEventHubsReceiver.scala:62]: Instead of adding a parameter, let's create a filed for the plugin as the following and remove all unnecessary changes.
  private lazy val metricPlugin: Option[MetricPlugin] = ehConf.metricPlugin()
  private lazy val client: EventHubClient = ClientConnectionPool.borrowClient(ehConf)
  private var receiver: PartitionReceiver = createReceiver(startSeqNo)

File: core/src/main/scala/org/apache/spark/eventhubs/client/CachedEventHubsReceiver.scala [hunk 11]
```diff
@@ -230,12 +227,17 @@ private[client] class CachedEventHubsReceiver private (ehConf: EventHubsConf,
...
+    }
+    assert(validateSize == batchCount)
 
     val elapsedTimeMs = TimeUnit.NANOSECONDS.toMillis(elapsedTimeNs)
     logInfo(s"(TID $taskId) Finished receiving for $nAndP, consumer group: ${ehConf.consumerGroup
       .getOrElse(DefaultConsumerGroup)}, batchSize: $batchSize, elapsed time: $elapsedTimeMs ms")
-
+    eventHubsReceiverListener.foreach(listener => {
+      listener.onReceiveMetric(nAndP, batchCount, batchSizeInBytes, elapsedTimeMs)
+    })
```
Comment (sjkwak) [inline core/src/main/scala/org/apache/spark/eventhubs/client/CachedEventHubsReceiver.scala:240]: You can revert all changes and add the following codes to invoke metric plugin object if it is defined.
    if (metricPlugin.isDefined) {
      val (_, batchSizeInBytes) =
        validate.map(eventData => (1, eventData.getBytes.length.toLong)).reduce {
          (countAndSize1, countAndSize2) =>
            (countAndSize1._1 + countAndSize2._1, countAndSize1._2 + countAndSize2._2)
        }
      metricPlugin.foreach(listener => {
        listener.onReceiveMetric(nAndP, batchCount, batchSizeInBytes, elapsedTimeMs)
      })
    }

File: core/src/main/scala/org/apache/spark/eventhubs/client/CachedEventHubsReceiver.scala [hunk 12]
```diff
@@ -274,23 +276,26 @@ private[spark] object CachedEventHubsReceiver extends CachedReceiver with Loggin
   private[eventhubs] override def receive(ehConf: EventHubsConf,
                                           nAndP: NameAndPartition,
                                           requestSeqNo: SequenceNumber,
-                                          batchSize: Int): Iterator[EventData] = {
+                                          batchSize: Int,
+                                          eventHubsReceiverListener: Option[MetricPlugin]
```
Comment (sjkwak) [inline core/src/main/scala/org/apache/spark/eventhubs/client/CachedEventHubsReceiver.scala:280]: you can remove it as not needed.

File: core/src/main/scala/org/apache/spark/eventhubs/client/CachedEventHubsReceiver.scala [hunk 13]
```diff
@@ -274,23 +276,26 @@ private[spark] object CachedEventHubsReceiver extends CachedReceiver with Loggin
...
+                                         ): Iterator[EventData] = {
     val taskId = EventHubsUtils.getTaskId
 
     logInfo(s"(TID $taskId) EventHubsCachedReceiver look up. For $nAndP, ${ehConf.consumerGroup
       .getOrElse(DefaultConsumerGroup)}. requestSeqNo: $requestSeqNo, batchSize: $batchSize")
     var receiver: CachedEventHubsReceiver = null
     receivers.synchronized {
       receiver = receivers.getOrElseUpdate(key(ehConf, nAndP), {
-        CachedEventHubsReceiver(ehConf, nAndP, requestSeqNo)
+        CachedEventHubsReceiver(ehConf, nAndP, requestSeqNo, eventHubsReceiverListener)
```
Comment (sjkwak) [inline core/src/main/scala/org/apache/spark/eventhubs/client/CachedEventHubsReceiver.scala:289]: you can remove it as not needed.

File: core/src/main/scala/org/apache/spark/eventhubs/client/CachedEventHubsReceiver.scala [hunk 14]
```diff
@@ -274,23 +276,26 @@ private[spark] object CachedEventHubsReceiver extends CachedReceiver with Loggin
...
     }
     receiver.receive(requestSeqNo, batchSize)
   }
 
   def apply(ehConf: EventHubsConf,
             nAndP: NameAndPartition,
-            startSeqNo: SequenceNumber): CachedEventHubsReceiver = {
-    new CachedEventHubsReceiver(ehConf, nAndP, startSeqNo)
+            startSeqNo: SequenceNumber,
+            eventHubsReceiverListener: Option[MetricPlugin] = None): CachedEventHubsReceiver = {
```
Comment (sjkwak) [inline core/src/main/scala/org/apache/spark/eventhubs/client/CachedEventHubsReceiver.scala:298]: you can remove it as not needed.

File: core/src/main/scala/org/apache/spark/eventhubs/rdd/EventHubsRDD.scala [hunk 15]
```diff
@@ -38,7 +38,8 @@ import org.apache.spark.{ Partition, SparkContext, TaskContext }
  */
 private[spark] class EventHubsRDD(sc: SparkContext,
                                   val ehConf: EventHubsConf,
-                                  val offsetRanges: Array[OffsetRange])
+                                  val offsetRanges: Array[OffsetRange],
```
Comment (sjkwak) [inline core/src/main/scala/org/apache/spark/eventhubs/rdd/EventHubsRDD.scala:41]: You can remove all changes in this file as not needed.

File: core/src/main/scala/org/apache/spark/eventhubs/utils/SimulatedCachedReceiver.scala [hunk 16]
```diff
@@ -42,7 +42,8 @@ private[spark] object SimulatedCachedReceiver extends CachedReceiver {
   override def receive(ehConf: EventHubsConf,
                        nAndP: NameAndPartition,
                        requestSeqNo: SequenceNumber,
-                       batchSize: Int): Iterator[EventData] = {
+                       batchSize: Int,
+                       eventHubsReceiverListener: Option[MetricPlugin] = None): Iterator[EventData] = {
```
Comment (sjkwak) [inline core/src/main/scala/org/apache/spark/eventhubs/utils/SimulatedCachedReceiver.scala:46]: you can rollback all changes in this file.

File: core/src/main/scala/org/apache/spark/sql/eventhubs/EventHubsForeachWriter.scala [hunk 17]
```diff
@@ -39,21 +39,38 @@ import org.apache.spark.sql.ForeachWriter
  */
 case class EventHubsForeachWriter(ehConf: EventHubsConf) extends ForeachWriter[String] {
   var client: EventHubClient = _
+  var totalMessageSizeInBytes = 0
```
Comment (sjkwak) [inline core/src/main/scala/org/apache/spark/sql/eventhubs/EventHubsForeachWriter.scala:42]: private lazy val metricPlugin: Option[MetricPlugin] = ehConf.metricPlugin()

File: core/src/main/scala/org/apache/spark/sql/eventhubs/EventHubsForeachWriter.scala [hunk 18]
```diff
@@ -39,21 +39,38 @@ import org.apache.spark.sql.ForeachWriter
...
     retryJava(client.send(event), "ForeachWriter")
+    totalMessageCount += 1
+    totalMessageSizeInBytes += event.getBytes.length
   }
 
   def close(errorOrNull: Throwable): Unit = {
     errorOrNull match {
-      case t: Throwable => throw t
+      case t: Throwable =>
+        ehConf.metricPlugin().foreach(_.onSendMetric(
```
Comment (sjkwak) [inline core/src/main/scala/org/apache/spark/sql/eventhubs/EventHubsForeachWriter.scala:62]: metricPlugin.foreach(..)

File: core/src/main/scala/org/apache/spark/sql/eventhubs/EventHubsForeachWriter.scala [hunk 19]
```diff
@@ -39,21 +39,38 @@ import org.apache.spark.sql.ForeachWriter
...
+          totalMessageCount,
+          totalMessageSizeInBytes,
+          System.nanoTime() - writerOpenTime,
+          isSuccess = false))
+        throw t
       case _ =>
+        ehConf.metricPlugin().foreach(_.onSendMetric(
```
Comment (sjkwak) [inline core/src/main/scala/org/apache/spark/sql/eventhubs/EventHubsForeachWriter.scala:69]: metricPlugin.foreach(..)

File: core/src/main/scala/org/apache/spark/sql/eventhubs/EventHubsWriteTask.scala [hunk 20]
```diff
@@ -43,6 +43,9 @@ private[eventhubs] class EventHubsWriteTask(parameters: Map[String, String],
 
   private val ehConf = EventHubsConf.toConf(parameters)
   private var sender: Client = _
+  var totalMessageSizeInBytes = 0
```
Comment (sjkwak) [inline core/src/main/scala/org/apache/spark/sql/eventhubs/EventHubsWriteTask.scala:46]: Same feedback as foreachwriter class.

File: core/src/main/scala/org/apache/spark/streaming/eventhubs/EventHubsDirectDStream.scala [hunk 21]
```diff
@@ -49,7 +50,8 @@ import org.apache.spark.streaming.scheduler.rate.RateEstimator
  */
 private[spark] class EventHubsDirectDStream private[spark] (_ssc: StreamingContext,
                                                             ehConf: EventHubsConf,
-                                                            clientFactory: EventHubsConf => Client)
+                                                            clientFactory: EventHubsConf => Client,
```
Comment (sjkwak) [inline core/src/main/scala/org/apache/spark/streaming/eventhubs/EventHubsDirectDStream.scala:53]: You can rollback all changes in this class as not needed.

File: core/src/test/scala/org/apache/spark/eventhubs/EventHubsConfSuite.scala [hunk 22]
```diff
@@ -20,12 +20,13 @@ package org.apache.spark.eventhubs
...
 
-import org.apache.spark.eventhubs.utils.EventHubsTestUtils
+import com.microsoft.azure.eventhubs.EventData
+import org.apache.spark.eventhubs.utils.{EventHubsTestUtils, MetricPlugin, MetricPluginMock}
 import org.json4s.NoTypeHints
 import org.json4s.jackson.Serialization
-import org.json4s.jackson.Serialization.{ read => sread }
-import org.json4s.jackson.Serialization.{ write => swrite }
-import org.scalatest.{ BeforeAndAfterAll, FunSuite }
+import org.json4s.jackson.Serialization.{read => sread}
```
Comment (sjkwak) [inline core/src/test/scala/org/apache/spark/eventhubs/EventHubsConfSuite.scala:27]: clean up all imports as they are not used.

File: core/src/main/scala/org/apache/spark/eventhubs/client/CachedEventHubsReceiver.scala (hunk 11 above)
Comment (sjkwak) [reply core/src/main/scala/org/apache/spark/eventhubs/client/CachedEventHubsReceiver.scala:240]: Here is updated version which addresses the issue of enumerating empty map.
    if (metricPlugin.isDefined) {
      val (validateSize, batchSizeInBytes) =
        validate.map(eventData => (1, eventData.getBytes.length.toLong)).reduce {
          (countAndSize1, countAndSize2) =>
            (countAndSize1._1 + countAndSize2._1, countAndSize1._2 + countAndSize2._2)
        }
      metricPlugin.foreach(listener => {
        listener.onReceiveMetric(nAndP, batchCount, batchSizeInBytes, elapsedTimeMs)

        assert(validateSize == batchCount)
      })
    } else {
      assert(validate.size == batchCount)
    }

File: core/src/main/scala/org/apache/spark/eventhubs/EventHubsConf.scala [hunk 23]
```diff
@@ -436,6 +437,18 @@ final class EventHubsConf private (private val connectionStr: String)
     set(MaxEventsPerTriggerKey, limit)
   }
 
+  def setMetricPlugin(metricPlugin: MetricPlugin): EventHubsConf = {
```
Comment (sjkwak) [inline core/src/main/scala/org/apache/spark/eventhubs/EventHubsConf.scala:440]: You also need to add "eventhubs.metricPlugin" to trimmed version of EventHubsConf - line 162.
